### DISCUSSION TITLE
UC UI Discussion

### DISCUSSION THREAD

**Discussion by dennyglee** [unitycatalog/unitycatalog#466] [Ideas]
The purpose of this discussion is a high level track on Unity Catalog UI discussions.

**Comment by JuanPabloDiaz** [comment]
I noticed the `uiux` label. Are there any `good first issues` related to the UI? I can work on something else while the BE is ready. #461

**Reply by dennyglee** [reply]
Not yet but only because I think we're pushing hard for the 0.2 release schedule.  As soon as we get this out, I think we'll able to start creating more `good first issues`.  Perhaps let's chat about this during next week's UI contributor meeting so we can plan accordingly?

**Reply by JuanPabloDiaz** [reply]
Looking forward to it!

**Comment by ognis1205** [comment]
I'm going to start working on [this issue](https://github.com/unitycatalog/unitycatalog/issues/431) from today. It seems that the current implementation mixes `export function` and `export const`, but is there any implementation policy regarding this? Personally, I'd prefer to use arrow functions unless there's a special reason not to.

**Reply by dennyglee** [reply]
Oh good point - @creechy , I would love to get your take on this approach?

**Reply by josephmock-db** [reply]
I'm not entirely familiar with `export function` and `export const`. is that in the OpenAPI spec or somewhere else? Can you explain a little more. (cc: @jamieknight-db , @vikrantpuppala )

**Reply by ognis1205** [reply]
Apologies for the delay in my response and the UC-related implementation. I've been a bit busy with other work and projects recently. The `export function` and `export const` are about arrow functions in TypeScript (ECMAScript). In the existing implementation, almost all component implementations follow the `export function` notation, but there are [a few exceptions](https://github.com/unitycatalog/unitycatalog/blob/35af13e412da6e9fa9d70232a4bfdfe37c982634/ui/src/utils/NotificationContext.tsx#L15) where arrow functions are used with `export const`. I thought it might be cleaner to have some sort of convention for this. It's a minor thing, but I think having consistency would make the code easier to read.

**Reply by ognis1205** [reply]
My preference is to use arrow functions, type only the function arguments, and let the transpiler infer the return type without specifying it.

**Reply by ognis1205** [reply]
Sorry for my ambiguity.

**Reply by creechy** [reply]
Got it... I'll defer to @yc-shawn and @jamieknight-db who have more TypeScript experience.

**Reply by jamieknight-db** [reply]
Hi @ognis1205! I am not overly opinionated on this, but my personal preference is to use `const` over function. I'll defer to @yc-shawn as well.

**Comment by ognis1205** [comment]
I think the use of [JSDoc](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) is worth considering.

**Comment by ognis1205** [comment]
@jamieknight-db @yc-shawn 

I don't consider myself the most suitable person, but once the OpenAPI PR is complete, could I propose some refactoring ideas (such as unifying file extensions to .ts/.tsx, standardizing function vs. arrow function, etc.)? I believe having some level of coding conventions would be beneficial as we bring new people into the project.

**Reply by dennyglee** [reply]
I think that makes sense @ognis1205  - do you think you could write up a simple design spec on what you think the refactoring should look like so we can share with the wider community?

**Reply by ognis1205** [reply]
Thank you for your response @dennyglee! I’ll summarize it in a Google document or something. Based on that, I’d like to gather input from other members as well.

**Comment by ognis1205** [comment]
Before making a proposal, verification is necessary, but I would also suggest adopting Vite as a build tool instead of CRA due to the lack of active development in CRA and the performance of its build and development server:

 - https://github.com/facebook/create-react-app
 - https://github.com/vitejs/vite

**Reply by dennyglee** [reply]
Hey @ognis1205 - I'm inclined to agree with you but would love to have others chime in on their thoughts.  Do you think there is a lot work involved with switching to Vite (vs. CRA) with our current UI code?

**Reply by ognis1205** [reply]
@dennyglee

Got it. I’ll include the Vite discussion in the above refactoring and coding guidelines proposal and document it. I don't think migrating to Vite will require too much effort, but let me look into it a bit more.

**Reply by ognis1205** [reply]
NOTE:

https://github.com/vitejs/vite/issues/2433

**Comment by ognis1205** [comment]
[Part I: Refactoring](https://docs.google.com/document/d/1s4O_mYouMcF1Pd3-MFsq6kHmqox4iNNQLh70dDxhNNw/edit?usp=sharing)
[Part II: Coding Convention](https://docs.google.com/document/d/1f9G7_Uc1fjw2Knit2XORvZvKA74au65mn2uQ78H4L3g/edit?usp=sharing)
[Part III: Vite Migration](https://docs.google.com/document/d/10gn1fiGDz9XselhWCP-w4oDzEkOGHhrTG3C6Wd9Dho8/edit?usp=sharing)

**Reply by ognis1205** [reply]
@jamieknight-db @yc-shawn @dennyglee 

Any plan to update dependency?

<img width="912" alt="Screenshot 2024-12-25 at 18 03 12" src="https://github.com/user-attachments/assets/d2e851a1-dfa0-4213-bcbc-b09a6446fdbe" />

**Reply by dennyglee** [reply]
Good call @ognis1205  - do you mind making a PR so we can do a quick test run and validate?

**Reply by ognis1205** [reply]
@dennyglee 

In the future, I believe we’ll need to consider using Dependabot, including its validation, but I’ll also prepare the creation of this PR. Thank you for the feedback.

**Reply by dennyglee** [reply]
@ognis1205 Completely fair - I think we do need to do this as well.  Let me add this to one of our issues and tag myself to enable this, eh?!

**Reply by ognis1205** [reply]
Thank you for your cooperation, @dennyglee !

@jamieknight-db @yc-shawn @dennyglee 
As we discussed earlier, I’ve put together a proposal regarding the refactoring plan and migration to Vite above. Please review it when you have time. I’ll also share it on Slack to gather others who might be interested in this topic.

**Reply by dennyglee** [reply]
Thanks @ognis1205  - could you give us comment access to the docs?

**Reply by ognis1205** [reply]
@dennyglee 

Sorry for not noticing. I granted comment permissions to everyone.

**Reply by yc-shawn** [reply]
@ognis1205 Thanks for investigating and created a plan, but given the current project side (which is small), the improvement might be marginal. Besides, if we do want to make refactor to the project, have we consider Next.js to wrap the entire project which I believe Next.js is one of the most popular framework for React nowadays.

**Reply by ognis1205** [reply]
@yc-shawn Thanks for the feedback!

Regarding refactoring, I see it more as part of maintaining code conventions rather than merely consolidating code. For example, while `utils/NotificationContext.tsx` and `context/auth-context.tsx` use the same design pattern, I personally feel that their file naming/structuring lacks consistency. From the perspective of refactoring for code consolidation, the benefits might be limited given that this is still a small project. However, it could still yield meaningful results, such as reducing the review workload and implicitly providing implementation guidelines for new developers.

Next.js is, of course, one of the framework options to consider. While CRA and Vite lean more towards being React build tools, I believe Next.js has stronger characteristics as a full-stack React framework. (For the same reason that the project is currently small, I didn’t include Next.js in this evaluation. Additionally, utilizing features like SSR would require a certain amount of modifications.) Therefore, I think the framework choice should be based on the following decision tree:

 - Need SSR/SSG, routing, and full-stack features?
    Go with Next.js.
 - Want blazing-fast development and a simpler toolchain?
    Choose Vite.

In any case, I personally feel that continuing to use CRA in its current state is not ideal. While there is an option to use Vite for development servers and Next.js for production builds, there may be some feature limitations with SSR or API routes. I can also perform validations for migrating to Next.js if needed. What do you think?

**Reply by yc-shawn** [reply]
@ognis1205 I don't think we need SSR at the moment, and fast dev & better toolchain is useful for sure. However, do you know if it’s straightforward to start with Vite and later introduce Next.js?

**Reply by ognis1205** [reply]
@yc-shawn 

That's an excellent point. In the long run, migrating to Next.js could offer significant benefits. Since there don’t appear to be any objections to this migration, I’d be happy to explore that plan further. Alternatively, I can test it in my local repository if that works for you. What are your thoughts on this approach?

**Comment by ognis1205** [comment]
Note:

https://github.com/okta/okta-signin-widget/issues/2131

**Comment by ognis1205** [comment]
@jamieknight-db @yc-shawn cc: @dennyglee 

Would it be possible for me to propose migrating to Next.js in my local repository? I believe using CRA may not be ideal. This might take some time since I have several other tasks to work on.

**Reply by dennyglee** [reply]
Hey @ognis1205  - to ensure I understand the context, the proposal here is to migrate your version of the UC UI into next.js in your own repo?  If so, yes I think this would be a good idea so this way we can showcase its viability, eh?!  I love this idea!  Please let me know if I'm misunderstanding the ask, eh?!

**Reply by ognis1205** [reply]
Thanks for the replay @dennyglee ,

Sorry for the ambiguous wording. What I meant was that I plan to create a migration testing branch on my unitycatalog fork and validate the migration there—just like I did for [the vitejs migration test](https://github.com/ognis1205/unitycatalog/tree/feat/use-vite). The UI I built a long time ago, i.e., my version of the UI, was originally implemented using Next.js. Would that be okay?

**Reply by ognis1205** [reply]
If migration itself is not desired, I am happy to conclude this discussion. Also, if my participation in the project would not be beneficial overall, I will refrain from sharing my opinions or making contributions.

**Reply by jamieknight-db** [reply]
@ognis1205 I'm happy to review a migration to Next.js - although I have not used it before so I'm not the best reviewer. My contributions will be limited moving forward as I've been moved to a different internal team (same one as @yc-shawn) so as long as you feel comfortable taking the lead on this and it's an effort you feel is valuable, then I say go for it 👍

**Reply by dennyglee** [reply]
Like @jamieknight-db, I definitely want to add my vote for definitely doing this @ognis1205 - really appreciate you pushing forward with this, eh?!

**Reply by JuanPabloDiaz** [reply]
I would like to contribute in case you need it @ognis1205.

**Reply by ognis1205** [reply]
> @ognis1205 I'm happy to review a migration to Next.js - although I have not used it before so I'm not the best reviewer. My contributions will be limited moving forward as I've been moved to a different internal team (same one as @yc-shawn) so as long as you feel comfortable taking the lead on this and it's an effort you feel is valuable, then I say go for it 👍

@jamieknight-db 

Whether someone is a good reviewer is not necessarily determined solely by their experience with a tool or framework, and I’ve been helped in many ways by past reviews. While it does make sense to consider experience when judging whether someone is a suitable reviewer, one of the great things about OSS is that it allows people to engage with unfamiliar technologies alongside others. So, I’d really appreciate it if you could continue reviewing whenever time allows.

**Reply by ognis1205** [reply]
@dennyglee @JuanPabloDiaz 

I'll give the migration task a try as well. I can either work on a PR together with Juan or assist him so that he can submit the PR himself. For now, let's take on the challenge and see how it goes.

**Reply by JuanPabloDiaz** [reply]
Sounds good to me!
